### ISSUE TITLE
Serenity/JS 3.0 - WDIO, how to set implicit wait?

### ISSUE THREAD

**Issue by viper3400** [serenity-js/serenity-js#1101]
How can I set the [implicit wait timeout for WDIO](https://webdriver.io/docs/timeouts/#session-implicit-wait-timeout)?

We've shortly discussed this in [Gitter](https://gitter.im/serenity-js/Contributors?at=61e5581f742c3d4b21a022c7) but it seems I cannot configure this and default timeout seems to be 0.

How to reproduce:
- I created a new branch based on the current main branch: https://github.com/viper3400/serenity-js/tree/features/async-ensure
- I copied 
  - the existing `PageElements.spec.ts` to create a `PageElements_async.spec.ts` 
  - and the `dynamic_shopping_list.html` to `dynamic_shopping_list_async.html`
- I created a little [timeout when clicking on the x in the shopping list](https://github.com/viper3400/serenity-js/blob/e312935244fb1fb639ab82fcfe12e20723b333d8/integration/web-specs/static/screenplay/models/page-elements/dynamic_shopping_list_async.html#L15)
- But even with the 500ms timeout [the test fails](https://github.com/viper3400/serenity-js/blob/e312935244fb1fb639ab82fcfe12e20723b333d8/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L97)

As workaround I would have to 
- either `Wait` for a duration before `Ensure.that` or `Check.whether`  an element is visible 
- or `Wait.until` the element is visible (instead of `Ensure`)

**Comment by jan-molak** [comment]
Based on what I see in the [WebdriverIO docs](https://webdriver.io/docs/autowait/#implicit-vs-explicit):

> The WebDriver protocol offers implicit timeouts that allow specify how long the driver is suppose to wait for an element to show up. By default this timeout is set to 0 and therefore makes the driver return with an no such element error immediately if an element could not be found on the page. Increasing this timeout using the [setTimeout](https://webdriver.io/docs/api/browser/setTimeout/) would make the driver wait and increases the chances that the element shows up eventually.

Now, [`setTimeout` docs](https://webdriver.io/docs/api/browser/setTimeout/) propose this approach:

```typescript
it('should change timeout duration for session with long code duration', async () => {
    await browser.setTimeout({
        'pageLoad': 10000,
        'script': 60000
    });
    // Execute code which takes a long time
    await browser.executeAsync((done) => {
        console.log('Wake me up before you go!');
        setTimeout(done, 59000);
    });
});
```
Based on the information provided in [WebdriverIO docs on timeouts](https://webdriver.io/docs/timeouts), we should be able to configure the implicit timeouts as follows:

```typescript
browser.setTimeout({ 'implicit': 5_000 });
```

The trouble with the above is that `browser.setTimeout` is an async call, so the below **won't work**:

```typescript
import { Actor, Cast } from '@serenity-js/core';
import { BrowseTheWebWithWebdriverIO } from '@serenity-js/webdriverio';

export class Actors implements Cast {
    prepare(actor: Actor): Actor {
        browser.setTimeout({ 'implicit': 5_000 });  // won't work, this returns a promise that has to be synchronised

        return actor.whoCan(
            BrowseTheWebWithWebdriverIO.using(browser),
        );
    }
}
```
But, we could take a different approach.

# Approach 1 - custom interaction

We could _extend_ `BrowseTheWebWithWebdriverIO`, and add a new method that sets the timeouts:
```typescript
export class MyBrowseTheWeb extends BrowseTheWebWithWebdriverIO {
  setImplicitTimeout(milliseconds: number): Promise<void> {
    return this.browser.setTimeout({ 'implicit': 5_000 });
  }
}
```
Then, create an interaction that invokes the above method:
```typescript
const SetImplicitTimeout = (milliseconds: number) =>
  Interaction.where(`#actor sets implicit timeout to ${ milliseconds }`, actor => {
    return (BrowseTheWeb.as(actor) as MyBrowseTheWeb).setImplicitTimeout(milliseconds);
  })
```

# Approach 2 - custom ability

Alternatively, we could come up with a version of `BrowseTheWebWithWebdriverIO` that sets the timeouts whenever the ability is [initialised](https://serenity-js.org/modules/core/file/src/screenplay/abilities/Initialisable.ts.html):

```typescript
import { BrowseTheWebWithWebdriverIO } from '@serenity-js/webdriverio';
import { Initilisable } from '@serenity-js/core';

export class MyBrowseTheWeb extends BrowseTheWebWithWebdriverIO implements Initilisable  {
  private initilised: boolean = false;

  initialise(): Promise<void> {
    return this.browser.setTimeout({ 'implicit': 5_000 }.then(() => {
      this.initialised = true;
    });
  }
  isInitilised(): boolean {
    return this.initilised;
  }
}
```

# Approach 3 - add to `@serenity-js/web`

Since this is something that could be useful to many other people, we could also consider adding a way to configure implicit timeouts to `BrowseTheWeb`. We'd just need to think about how to make it consistent across the different drivers.

**Comment by viper3400** [comment]
I think all approaches have their pros & cons. The first thing wich came into my mind when reading your suggestions was, what the “lifetime” of this implicit timeout would be.

So lets pinpoint some use cases:

When using WebdriverIO the default implicit timeout is set to 0. I want to be able to set & configure this implicit timeout
- For all of my tests and specification files globally
- Just for one test or specification
- Just for one task or one interaction

Let me start with **approach 2**: 

This would mean, whenever I create an actor I would be able to set a implict timeout. This value will be tied to this actor and the lifetime of this actor.

I would be able to define a global `Actor` with a global timeout or an actor just for one spec file or even for one `Task`.

But is it really an ability of an actor “A” to wait, lets say 5 seconds, until they check an element and then the ability of an actor “B” to wait 10 seconds? 

Agreed, it could be the patience of the user. ;-)

So, **aproach 1**:
What would that mean? I would be able to set a custom timeout in any task using this interaction at any time? What would be the lifetime of this timeout? When I set the value to 5000, when would it be set back? The next time I call the interaction? When the next spec starts?

This seems to me a bit like `UseAngular.enableSynchronisation` and `UseAngular.disableSychnronisation` and to be honest, I’m not sure what the lifetime with this is either. I assume it ends with the spec?

**Approach 3** seems to be a question of the layer of abstraction? I think, as there are dynamic and animated websites are in the world, ever browser automation tool has to deal with this problem. But I also think, they have to deal on their own and they may do in different ways. So I first thing I thought, was that this should be something specific for the used tool.

But not having it in `@serenity-js/web` would mean you’d have to import the `Interaction` or the `Abilitiy` from a `@serenity-js/webdriverio` and then your code would probably have dependencies to an automation tool in places you don’t want it (and you’re back in Serenity/JS 2)

**Comment by jan-molak** [comment]
I think the question of scope and granularity is a good one.


It feels to me like we have several levels here:
- global level - I want to set timeouts for all my tests, no matter what websites or APIs they touch
- ability level - timeouts are specific to any operations performed using a given ability, used by a given actor (I think the same pattern could apply to `CallAnApi`, btw)
- interaction level - timeouts are specific to a particular interaction, i.e. `PageElement.located(By.css('...')).awaitedFor(Duration.ofSeconds(5))`, `Navigate.to('...').awaitedFor(Duration.ofSeconds(3))` (not sure if `awaitedFor` is the right name, probably not)

I'd definitely prefer to have a consistent mechanism here to avoid driver-specific implementation leakage.

> But is it really an ability of an actor “A” to wait, lets say 5 seconds, until they check an element and then the ability of an actor “B” to wait 10 seconds?

Ha, it would be interesting to see how this could work; a "patient actor", vs an "impatient actor". I could see that being useful in performance benchmarking - for example, an impatient actor would wait for up to 500ms for a page to load, while a patient one would work for 5s; Then you run the tests and see if all actors' needs are satisfied.

**Comment by viper3400** [comment]
Hi Jan. I figured it now  out completely, but now I found that there _is_ a timeout when just ensuring the visibility of a “simple” page element. Seems the problem is somewhere in the enhanced “.where” pattern. I haven’t reproduced this, yet but may do this, if you don’t see something obvious here?

**Comment by jan-molak** [comment]
Hey @viper3400

> if you don’t see something obvious here

Hmm, not sure what bit you're referring to? Timeouts in `.where`?

**Comment by viper3400** [comment]
Exactly. But it is just an impression that their is a kind “implicit wait” when using a straight forward `PageElement.located(By.css’…’)` but not when you would select `PageElements.located(By.css’…’).where(Text.of(x), equals(y)`.

But I can try to adjust my scenario to get some reliable informations about this behaviour. 
My point is, that this may not be a global issue at all. But lets see.

**Comment by jan-molak** [comment]
Cool, thanks for investigating. Please keep me posted!

**Comment by viper3400** [comment]
- we have [a page element that not exists](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L69)
- we have

**Comment by viper3400** [comment]
Hi @jan-molak , I was able to reproduce the issue which left me a bit confused the last time. There is a different behaviour if you:

1. Ensure that a certain not existing page element is not visible and 
2. Ensure that the child of this not existing page element is not visible

- we have [a page element that not exists](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L69)
- we have [a child of this page element](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L73)

Now, when just ensuring that [the parent element](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L103) is not visible, the test will fail immediately.

When ensuring, that [the child element of this parent page element](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/web-specs/spec/screenplay/models/PageElements_async.spec.ts#L109) is not visible, WDIO will wait for the [waitForTimeout](https://github.com/viper3400/serenity-js/blob/262a39aee44f6307b3ce24a6a5e9c8c1472405e2/integration/webdriverio-web/wdio.conf.ts#L104).

You should raise the log level to debug and you should also raise the `waitForTimeout` to get better insights into to this.

Some words to my real world use case: I use this strategy to ensure that special roles do not have access to some icons in a toolbar when they have no permissions to do so. The toolbar is the container for the icons (the parent). If a role has no permissions for any of the icons, the whole container won’t be present.

So not sure, if this is worth another issue as this is a bit different to the original issue in #1101 ?

**Comment by jan-molak** [comment]
@viper3400 - do you think we could close this ticket now that we have #1083 and #1522 in place? Or is it still relevant?

**Comment by viper3400** [comment]
@jan-molak, sorry for being silent so long. Just give me a few days to follow up this issue.

**Comment by viper3400** [comment]
Hi @jan-molak, this issue looks like a little bit of everything without a clear idea about the actions to take. The effects described here are mitigated by the issues you've mentioned and I like to ad #1339 to this list.

I raised this issue at the time when we switched from Protractor to WebdriverIO under the hood and had to fight with flaky tests.

What we learned then: It's better you know how your page under test behaves in time. This is, `Wait.until` an element that should appear / disappear async is (not) visible. Just `Ensure` an element, if you're sure it should (not) be visible at a point in time.

(This is why I'm a bit sceptical about the `Ensure.eventually`. But maybe I change my mind, once I've tried it myself 😏)

Anyway, let's close this issue as we don't have issues at the moment in this context.
